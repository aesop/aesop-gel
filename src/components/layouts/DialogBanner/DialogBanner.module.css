@import 'styles/variables/index';
@import 'styles/mixins/index';

.base {
  position: fixed;
  z-index: var(--z-index-fixed-dialog-banner);
  right: 0;
  left: 0;
  padding: 1.5rem rem(40px);

  @media (--viewport-sm-only) {
    padding: 1.5rem 1.5rem 2.2rem 1.5rem;
  }

  @media (--viewport-md) {
    display: grid;
    grid-template-columns: 15% 63% 22%;
  }

  @media (--viewport-lg) {
    grid-template-columns: 20% 60% 20%;
  }
}

.top {
  top: 0;
}

.bottom {
  @extend %slideOutDown;

  bottom: 0;
}

/* @TODO Handle bottom prop */
.isVisible {
  &.bottom {
    @extend %slideInUp;
  }
}

.dark {
  background-color: var(--color-secondary);
}

.black {
  background-color: var(--color-grey-10);
}

.contentContainer {
  @media (--viewport-md) {
    padding: 0 rem(24px);
  }
}

.title {
  color: var(--color-white);

  @media (--viewport-sm-only) {
    margin-top: rem(28px);
    margin-bottom: rem(10px);
  }
}

.message {
  color: var(--color-grey-65);
  font-size: rem(15px);

  @media (--viewport-sm-only) {
    &:last-of-type {
      margin-bottom: rem(14px);
    }
  }
}

.warningIcon {
  fill: #999;
}

.closeButton {
  position: absolute;
  top: rem(1.75rem);
  right: rem(2.5rem);

  @media (--viewport-sm-only) {
    top: rem(1.5rem);
    right: rem(1.5rem);
  }

  &:hover,
  &:focus {
    svg {
      fill: white;
    }
  }
}

.closeIcon {
  fill: #999;
}

.closeContainer {
  @media (--viewport-md) {
    align-self: flex-end;
    text-align: right;
  }
}

.ctaButton {
  font-size: rem(15px);
}
