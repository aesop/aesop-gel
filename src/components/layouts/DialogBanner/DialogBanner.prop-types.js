import PropTypes from 'prop-types';
import { getPropTypeOptionValues } from '~/utils/propTypes/PropTypes';

export const POSITION_PROP_TYPE_OPTIONS = {
  TOP: 'top',
  BOTTOM: 'bottom',
};

export const THEME_PROP_TYPE_OPTIONS = {
  DARK: 'dark',
  BLACK: 'black',
};

export const PROP_TYPES = {
  className: PropTypes.string,
  cta: PropTypes.object, // @TODO make hyperlink proptype
  handleClose: PropTypes.func.isRequired,
  isVisible: PropTypes.boolean,
  message: PropTypes.string.isRequired,
  position: PropTypes.oneOf(
    getPropTypeOptionValues(POSITION_PROP_TYPE_OPTIONS),
  ),
  showCloseButton: PropTypes.boolean,
  theme: PropTypes.oneOf(getPropTypeOptionValues(THEME_PROP_TYPE_OPTIONS)),
  title: PropTypes.string.isRequired,
};

export default PROP_TYPES;
