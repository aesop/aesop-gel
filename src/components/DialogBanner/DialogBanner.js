import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { useEscapeKeyListener } from '~/customHooks/useEscapeKeyListener';
import { useOverflowHidden } from '~/customHooks/useOverflowHidden';
import Button from '~/components/Button';
import Heading from '~/components/Heading';
import Hyperlink from '~/components/Hyperlink';
import Icon from '~/components/Icon';
import { P } from '~/components/Paragraph';
import Overlay from '~/components/Overlay';
import Transition from '~/components/Transition';
import styles from './DialogBanner.module.css';

const DialogBanner = ({
  className,
  copy,
  cta,
  hasCloseButton,
  heading,
  isVisible,
  message,
  onClose,
  position,
  theme,
}) => {
  useEscapeKeyListener(onClose);
  useOverflowHidden(isVisible);

  const classSet = cx(
    styles.base,
    { [styles.isVisible]: isVisible },
    styles[theme],
    styles[position],
    className,
  );

  return (
    <>
      <Overlay isVisible={isVisible} onClose={onClose} />
      <Transition
        hasCSSTransitionMountOnEnter={true}
        hasCSSTransitionUnmountOnExit={true}
        isActive={isVisible}
        type="slideDown"
      >
        <aside className={classSet}>
          <div className={styles.iconContainer}>
            <Icon
              className={styles.warningIcon}
              height={22}
              name="warning"
              width={22}
            />
          </div>
          <div className={styles.contentContainer}>
            <Heading
              className={styles.heading}
              level="1"
              noTopMargin={true}
              size="large"
            >
              {heading}
            </Heading>
            <P className={styles.message}>{message}</P>
          </div>
          <div className={styles.closeContainer}>
            {hasCloseButton && (
              <Button
                className={styles.closeButton}
                isInline={true}
                onClick={onClose}
                title={copy.closeButtonTitle}
              >
                <Icon
                  className={styles.closeIcon}
                  height={12}
                  name="close"
                  width={12}
                />
              </Button>
            )}
            {cta && cta.text && (
              <Hyperlink
                className={styles.ctaButton}
                dataTestRef={cta.dataTestRef}
                id={cta.id}
                isInline={true}
                target={cta.openInANewWindow ? '_blank' : '_self'}
                theme={'light'}
                title={cta.title}
                url={cta.url}
              >
                {cta.text}
              </Hyperlink>
            )}
          </div>
        </aside>
      </Transition>
    </>
  );
};

DialogBanner.propTypes = {
  className: PropTypes.string,
  copy: PropTypes.shape({
    closeButtonTitle: PropTypes.string,
  }),
  cta: PropTypes.object,
  hasCloseButton: PropTypes.bool,
  isVisible: PropTypes.bool,
  message: PropTypes.string.isRequired,
  onClose: PropTypes.func.isRequired,
  position: PropTypes.oneOf(['top', 'bottom']),
  theme: PropTypes.oneOf(['dark', 'black']),
  heading: PropTypes.string.isRequired,
};

DialogBanner.defaultProps = {
  className: undefined,
  copy: {
    closeButtonTitle: 'Close Dialog Banner',
  },
  cta: undefined,
  hasCloseButton: true,
  isVisible: false,
  message: undefined,
  onClose: undefined,
  position: 'top',
  theme: 'dark',
  heading: undefined,
};

export default DialogBanner;
