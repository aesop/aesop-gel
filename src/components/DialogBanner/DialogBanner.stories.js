import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, text, select } from '@storybook/addon-knobs';
import { State } from '~.storybook/State';
import '~.storybook/storybook.module.css';
import Button from '~/components/Button';
import DialogBanner from './DialogBanner';

storiesOf('Components/DialogBanner', module).add('Base component', () => (
  <State initialState={{ isVisible: false }}>
    {(state, setState) => (
      <>
        <DialogBanner
          cta={{
            text: 'Need assistance?',
            url: '/contact-us/',
          }}
          isVisible={state.isVisible}
          message={text(
            'message',
            'Our apologies, You have exceeded the maximum allowed quantity of this item.',
          )}
          onClose={() => setState({ ...state, isVisible: false })}
          // @ts-ignore
          position={select('position', ['top', 'bottom'], 'top')}
          showCloseButton={boolean('showCloseButton', true)}
          // @ts-ignore
          theme={select('theme', ['dark', 'black'], 'dark')}
          title={text('title', 'Maximum allowed quantity exceeded.')}
        />
        <Button
          onClick={() => setState({ ...state, isVisible: true })}
          title="open Dialog Banner"
        >
          Open Dialog Banner
        </Button>
      </>
    )}
  </State>
));
