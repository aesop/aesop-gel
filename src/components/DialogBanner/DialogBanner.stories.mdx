import { Fragment } from 'react';
import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { withKnobs } from '@storybook/addon-knobs';
import { State } from '~.storybook/State';
import { object, text, select, boolean } from '@storybook/addon-knobs';
import Button from '~/components/Button';
import DialogBanner from './DialogBanner';

<Meta
  title="Components/DialogBanner"
  component={DialogBanner}
  decorators={[withKnobs]}
/>

# DialogBanner

<Preview>
  <Story name="Base component">
    <State initialState={{ isVisible: false }}>
      {(state, setState) => (
        <Fragment>
          <DialogBanner
            cta={{
              text: 'Need assistance?',
              url: '/contact-us/',
            }}
            heading={text('heading', 'Maximum allowed quantity exceeded.')}
            isVisible={state.isVisible}
            message={text(
              'message',
              'Our apologies, You have exceeded the maximum allowed quantity of this item.',
            )}
            onClose={() => setState({ ...state, isVisible: false })}
            position={select('position', ['top', 'bottom'], 'top')}
            showCloseButton={boolean('showCloseButton', true)}
            theme={select('theme', ['dark', 'black'], 'dark')}
          />
          <Button
            onClick={() => setState({ ...state, isVisible: true })}
            title="open Dialog Banner"
          >
            Open Dialog Banner
          </Button>
        </Fragment>
      )}
    </State>
  </Story>
</Preview>

## API

<Props of={DialogBanner} />

## Usage Example

```
<DialogBanner
  cta={{
    text: 'Need assistance?',
    url: '/contact-us/',
  }}
  heading="..."
  isVisible={isVisible}
  message="..."
  onClose={callback}
/>
```
