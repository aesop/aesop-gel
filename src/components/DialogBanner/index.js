/* istanbul ignore file */
import { DialogBanner } from './DialogBanner';

export { DialogBanner };
