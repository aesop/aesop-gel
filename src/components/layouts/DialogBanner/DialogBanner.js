import React from 'react';
import cx from 'classnames';
import { useEscapeKeyListener } from '~/hooks/useEscapeKeyListener';
import { useOverflowHidden } from '~/hooks/useOverflowHidden';
import Button from '~/components/elements/Button';
import Heading from '~/components/elements/Heading';
import Icon from '~/components/elements/Icon';
import Overlay from '~/components/elements/Overlay';
import P from '~/components/elements/Paragraph';
import Transition from '~/components/elements/Transition';
import { PROP_TYPES, PROP_TYPE_OPTIONS } from './DialogBanner.prop-types';
import styles from './DialogBanner.module.css';

const DialogBanner = ({
  className,
  cta,
  handleClose,
  isVisible = false,
  message,
  position = PROP_TYPE_OPTIONS.POSITION.TOP,
  showCloseButton = true,
  theme = PROP_TYPE_OPTIONS.THEME.DARK,
  title,
}) => {
  useEscapeKeyListener(handleClose);
  useOverflowHidden(isVisible);

  const classSet = cx(
    styles.base,
    { [styles.isVisible]: isVisible },
    styles[theme],
    styles[position],
    className,
  );

  return (
    <>
      <Overlay handleClose={handleClose} isVisible={isVisible} />
      <Transition
        active={!!isVisible}
        mountOnEnter={true}
        type="slideDown"
        unmountOnExit={true}
      >
        <aside className={classSet}>
          <div className={styles.iconContainer}>
            <Icon
              className={styles.warningIcon}
              height={22}
              name="warning"
              width={22}
            />
          </div>
          <div className={styles.contentContainer}>
            <Heading
              className={styles.title}
              level="1"
              noTopMargin={true}
              size="large"
            >
              {title}
            </Heading>
            <P className={styles.message}>{message}</P>
          </div>
          <div className={styles.closeContainer}>
            {showCloseButton && (
              <Button
                className={styles.closeButton}
                inline={true}
                onClick={handleClose}
                title="Close dialog button"
              >
                <Icon
                  className={styles.closeIcon}
                  height={12}
                  name="close"
                  width={12}
                />
              </Button>
            )}
            {cta && cta.text && (
              <Button
                className={styles.ctaButton}
                dataTestRef={cta.dataTestRef}
                hasLightText={true}
                id={cta.id}
                inline={true}
                target={cta.openInANewWindow ? '_blank' : '_self'}
                title={`contact us link to ${cta.url}`}
                to={cta.url}
              >
                {cta.text}
              </Button>
            )}
          </div>
        </aside>
      </Transition>
    </>
  );
};

DialogBanner.propTypes = PROP_TYPES;

export default DialogBanner;
