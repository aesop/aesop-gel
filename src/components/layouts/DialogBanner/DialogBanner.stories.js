import React from 'react';
import { State, Store } from '@sambego/storybook-state';
import { storiesOf } from '@storybook/react';
import { boolean, text, select } from '@storybook/addon-knobs';
import 'normalize.css';
import '~/styles/base.module.css';
import '~/styles/storybook.module.css';
import Button from '~/components/elements/Button';
import DialogBanner from './DialogBanner';

const store = new Store({
  isVisible: false,
});

storiesOf('Layouts.DialogBanner', module).add('Base component', () => (
  <State store={store}>
    <DialogBanner
      cta={{
        text: 'Need assistance?',
        url: '/contact-us/',
      }}
      handleClose={() => store.set({ isVisible: false })}
      isVisible={store.get('isVisible')}
      message={text(
        'message',
        'Our apologies, You have exceeded the maximum allowed quantity of this item.',
      )}
      // @ts-ignore
      position={select('position', ['top', 'bottom'], 'top')}
      showCloseButton={boolean('showCloseButton', true)}
      // @ts-ignore
      theme={select('theme', ['dark', 'black'], 'dark')}
      title={text('title', 'Maximum allowed quantity exceeded.')}
    />
    <Button
      onClick={() => store.set({ isVisible: true })}
      title="open Dialog Banner"
    >
      Open Dialog Banner
    </Button>
  </State>
));
