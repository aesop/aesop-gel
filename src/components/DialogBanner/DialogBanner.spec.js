import React from 'react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import DialogBanner from './DialogBanner';

configure({ adapter: new Adapter() });

const mockFn = jest.fn();

describe('<DialogBanner />', () => {
  it('should be defined', () => {
    expect(DialogBanner).toBeDefined();
  });

  it('renders base component correctly', () => {
    const tree = renderer
      .create(
        <DialogBanner
          cta={{
            text: 'Need assistance?',
            url: '/contact-us/',
          }}
          hasCloseButton={true}
          heading="Maximum allowed quantity exceeded."
          isVisible={true}
          message="Our apologies, You have exceeded the maximum allowed quantity of this item."
          onClose={mockFn}
          position="top"
          theme="dark"
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
