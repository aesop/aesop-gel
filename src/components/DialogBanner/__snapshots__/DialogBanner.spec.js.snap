// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DialogBanner /> renders base component correctly 1`] = `
Array [
  <div
    className="base isVisible"
    onClick={[MockFunction]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={-1}
  />,
  <aside
    className="base isVisible dark top"
  >
    <div
      className="iconContainer"
    >
      <svg
        className="base warning light warningIcon"
        data-ref=""
        focusable="false"
        height={22}
        role="img"
        style={
          Object {
            "height": "22px",
            "width": "22px",
          }
        }
        tabIndex={-1}
        viewBox="0 0 50 50"
        width={22}
      >
        <g>
          <path
            d="M42.7,7.2C37.7,2.2,31.5,0,25,0S12.3,2.2,7.3,7.2c-9.7,9.7-9.7,25.4,0,35.2c5,5,11.2,7.2,17.7,7.2s12.7-2.5,17.7-7.2 C52.4,32.6,52.4,16.9,42.7,7.2z M41,40.6c-4.2,4.2-10,6.5-16,6.5s-11.7-2.2-16-6.5c-8.7-8.7-8.7-23,0-31.7c4.3-4.2,10-6.5,16-6.5 s11.7,2.3,16,6.5C49.7,17.6,49.7,31.9,41,40.6z"
          />
          <path
            d="M27,13.1V30h-4V13.1H27z M23,36.5h3.9v-4.3H23V36.5z"
          />
        </g>
      </svg>
    </div>
    <div
      className="contentContainer"
    >
      <h1
        className="base large dark heading"
      >
        Maximum allowed quantity exceeded.
      </h1>
      <p
        className="base dark message"
      >
        Our apologies, You have exceeded the maximum allowed quantity of this item.
      </p>
    </div>
    <div
      className="closeContainer"
    >
      <button
        className="base inlineStyle dark closeButton"
        disabled={false}
        onClick={[MockFunction]}
        title="Close Dialog Banner"
        type="button"
      >
        <svg
          className="base close light closeIcon"
          data-ref=""
          focusable="false"
          height={12}
          role="img"
          style={
            Object {
              "height": "12px",
              "width": "12px",
            }
          }
          tabIndex={-1}
          viewBox="0 0 50 50"
          width={12}
        >
          <g>
            <polygon
              className="close"
              points="50,5 45,0 25,20 5,0 0,5 20,25 0,45 5,50 25,30 45,50 50,45 30,25"
            />
          </g>
        </svg>
      </button>
      <a
        className="base left light ctaButton"
        href="/contact-us/"
        target="_self"
      >
        Need assistance?
      </a>
    </div>
  </aside>,
]
`;
